cucumber-picocontainer/src/main/java/io/cucumber/picocontainer/PicoFactory.java

```diff
@@ -31,8 +35,29 @@ public void start() {
                     .withCaching()
                     .withLifecycle()
                     .build();
+            Set<Class<?>> providers = new HashSet<>();
+            Set<Class<?>> providedClasses = new HashSet<>();
             for (Class<?> clazz : classes) {
-                pico.addComponent(clazz);
+                if (isProviderAdapter(clazz)) {
+                    providers.add(clazz);
+                    Class<?> providedClass = addProviderAdapter(clazz);
+                    providedClasses.add(providedClass);
+                } else if (isProvider(clazz)) {
+                    providers.add(clazz);
+                    Class<?> providedClass = addProvider(clazz);
+                    providedClasses.add(providedClass);
+                }
+            }
+            for (Class<?> clazz : classes) {
+                // do not add the classes that represent a picocontainer
+                // ProviderAdapter/Provider, and also do not add those raw
+                // classes that are already provided (otherwise this causes
+                // exceptional situations, e.g. PicoCompositionException
+                // with message "Duplicate Keys not allowed. Duplicate for
+                // 'class XXX'")
+                if (!providers.contains(clazz) && !providedClasses.contains(clazz)) {
```

Comment:
Logically okay. But please clean this up. There should be something in the streaming API that allows grouping by.